src/libecalc/dto/component_graph.py

```diff
@@ -1,15 +1,98 @@
 from __future__ import annotations
 
-from typing import List
+import abc
+from typing import Dict, List, Optional, Union
 
 from libecalc.common.component_info.component_level import ComponentLevel
 from libecalc.common.component_type import ComponentType
-from libecalc.common.graph import Graph, NodeID
+from libecalc.common.graph import Graph
+from libecalc.common.time_utils import Period
+from libecalc.common.utils.rates import TimeSeriesStreamDayRate
+from libecalc.core.result import EcalcModelResult
+from libecalc.core.result.emission import EmissionResult
 from libecalc.dto.node_info import NodeInfo
 
+ComponentID = str
 
-class ComponentGraph(Graph):
-    def get_parent_installation_id(self, node_id: NodeID) -> NodeID:
+
+class Component(abc.ABC):
+    """
+    Common info for a component in the model
+    """
+
+    @property
+    @abc.abstractmethod
+    def id(self) -> ComponentID: ...
+
+    @property
+    @abc.abstractmethod
+    def name(self) -> str: ...
+
+    @abc.abstractmethod
+    def get_node_info(self) -> NodeInfo: ...
+
+    @abc.abstractmethod
+    def get_ecalc_model_result(self) -> Optional[EcalcModelResult]:
+        """
+        TODO: We should look into removing this generic result, as it does not makes sense to provide the same result for all types of components.
+        """
+        ...
+
+
+class PowerProvider(Component, abc.ABC):
+    """
+    Provider that can deliver a power requirement
+    """
+
+    @abc.abstractmethod
+    def provide_power(self, power_requirement: List[TimeSeriesStreamDayRate]): ...
+
+
+class Emitter(Component, abc.ABC):
+    """
+    Something that causes emissions
+    """
+
+    @abc.abstractmethod
+    def get_emissions(self, period: Period = None) -> Dict[str, EmissionResult]: ...
+
+
+class PowerConsumer(Component, abc.ABC):
+    """
+    A consumer with a power requirement
+    """
+
+    @abc.abstractmethod
+    def get_power_requirement(self, period: Period = None) -> TimeSeriesStreamDayRate: ...
+
+
+class FuelConsumer(Emitter, abc.ABC):
+    """
+    An abstraction for provider and consumer that isn't able to give power requirement
+    """
+
+    @abc.abstractmethod
+    def get_fuel_usage(self, period: Period = None) -> TimeSeriesStreamDayRate: ...
+
+
+class ProcessGraph:
+    """
+    A set of components that should be evaluated together, but reported as separate components.
+    """
+
+    @abc.abstractmethod
+    def get_components(self) -> List[Component]: ...
+
+    @abc.abstractmethod
+    def evaluate(self) -> None: ...
+
+
+# TODO: Refactor ComponentGraph to EnergyGraph, protocol and composition instead of inherit from graph.
+#       EnergyGraph should know about ProcessGraph.
+
```

Comment:
This should be discussed. This is the interfaces we implement for different components. Currently asset and installation don't implement an interface, so the ComponentGraph is a bit inconsistent. We should figure out which types of components we have, or implement an alternative interface.

Reply:
good point. At the same time asset and installation is currently only "containers", and not a part of a network per se. However wrt. power networks and inter-networks they should play a role. I think it makes sense to focus on what is "below installation" here, as the installation and asset currently just "aggregates" the data ...